File: src/gestorAplicacion/personas/Paciente.java
/* Autores: Diego Andres Gracia Granados, Daniel Giraldo Vanegas, Elian David Velandia Riveros, Juan Camilo Gutierrez Martinez
y Santiago Arboleda Acevedo */
package gestorAplicacion.personas;

import gestorAplicacion.administracion.*;
import gestorAplicacion.servicios.*;

import java.util.ArrayList;
import java.util.Objects;

//Clase destinada a crear pacientes
public class Paciente extends Persona implements Pago{

    // Atributos
	private final HistoriaClinica historiaClinica;
    private CategoriaHabitacion categoriaHabitacion;
    private Habitacion habitacionAsignada;

    // Constructores
    public Paciente(int cedula, String nombre, String tipoEps,CategoriaHabitacion categoriaHabitacion) {
        super(cedula,nombre,tipoEps);
        this.categoriaHabitacion=categoriaHabitacion;
        this.historiaClinica = new HistoriaClinica(this);
    }
    // Sobrecarga de constructor
    public Paciente(int cedula, String nombre, String tipoEps){
        this(cedula, nombre, tipoEps, null);
    }

    // Métodos

    /* Método que se encarga de buscar los medicamentos disponibles y posteriormente busca los medicamentos
    que traten la enfermedad ingresada como parámetro */
    public ArrayList<Medicamento> medEnfermedad(Enfermedad enfermedad, Hospital hospital) {
        ArrayList<Medicamento> medicamentos = hospital.medicamentosDisponibles();
        ArrayList<Medicamento> medEnfermedades = new ArrayList<Medicamento>();
        for (Medicamento med : medicamentos){
            if (enfermedad.getNombre().equals(med.getEnfermedad().getNombre()) && enfermedad.getTipologia().equals(med.getEnfermedad().getTipologia())){
                medEnfermedades.add(med);
            }

        }
        return medEnfermedades;
    }

    /* Método que busca los doctores por especialidad y por el tipo de eps del paciente */
    public ArrayList<Doctor> buscarDoctorEps(String especialidad, Hospital hospital){
    	ArrayList<Doctor> doctoresPorEspecialidad = hospital.buscarTipoDoctor(especialidad);
    	ArrayList<Doctor> doctoresDisponibles = new ArrayList<Doctor>();

    	for(int i=1; i<=doctoresPorEspecialidad.size(); i++) {
    		if (Objects.equals(doctoresPorEspecialidad.get(i-1).getTipoEps(), getTipoEps())) {
    			doctoresDisponibles.add(doctoresPorEspecialidad.get(i-1));
    		}
    }
		return doctoresDisponibles;
  }

  /* Método que agrega una cita a la lista de citas de la historia clínica del paciente */
    public void actualizarHistorialCitas(Cita citaAsignada) {
    	historiaClinica.getHistorialCitas().add(citaAsignada);
    }

    //Sobre carga de métodos con calcular precio de los distintos servicios
    @Override
    public double calcularPrecio(Formula formula){
        double precio = 0;
        for (Medicamento med : formula.getListaMedicamentos()){
            if (formula.getPaciente().getTipoEps().equals("Contributivo")){
                precio += med.getPrecio()*0.8;
            }
            if (formula.getPaciente().getTipoEps().equals("Subsidiado")){
                precio += med.getPrecio()*0.7;
            }
            if (formula.getPaciente().getTipoEps().equals("Particular")){
                precio += med.getPrecio();
            }
        }
        return precio*(1+IVA);
    }

    @Override
    public double calcularPrecio(Cita citaAsignada) {

    	double precioTotal = 0;

    	if (Objects.equals(citaAsignada.getDoctor().getEspecialidad(),"General")) {
    		precioTotal += 5000;
    	}

    	if (Objects.equals(citaAsignada.getDoctor().getEspecialidad(),"Oftalmologia")) {
    		precioTotal += 7000;
    	}

    	if (Objects.equals(citaAsignada.getDoctor().getEspecialidad(),"Odontologia")) {
    		precioTotal += 10000;
    	}

    	if (Objects.equals(citaAsignada.getPaciente().getTipoEps(),"Contributivo")) {
    		precioTotal += 2000;
    	}

    	if (Objects.equals(citaAsignada.getPaciente().getTipoEps(),"Subsidiado")) {
    		precioTotal += 500;
    	}

    	if (Objects.equals(citaAsignada.getPaciente().getTipoEps(),"Particular")) {
    		precioTotal += 10000;
    	}

    	precioTotal = precioTotal*(1+IVA);
    	return precioTotal;
    }

    @Override
    public double calcularPrecio(CitaVacuna citaAsignada){
        double precioTotal=citaAsignada.getVacuna().getPrecio();

        if(Objects.equals(citaAsignada.getVacuna().getTipo(), "Obligatoria")){
            precioTotal += 1000;
        }
        if(Objects.equals(citaAsignada.getVacuna().getTipo(), "No obligatoria")){
            precioTotal += 3000;
        }
        if (Objects.equals(citaAsignada.getPaciente().getTipoEps(), "Contributivo")){
            precioTotal += 2000;
        }
        if (Objects.equals(citaAsignada.getPaciente().getTipoEps(), "Subsidiado")){
            precioTotal += 500;
        }
        if (Objects.equals(citaAsignada.getPaciente().getTipoEps(), "Particular")){
            precioTotal += 10000;
        }
        precioTotal=precioTotal*(1+IVA);
        return precioTotal;
    }

    @Override
    public  double calcularPrecio(Habitacion habitacionAsignada)
    {
        double precio=0;
        if (Objects.equals(getTipoEps(), "Subsidiado"))
        {
            precio=habitacionAsignada.getCategoria().getValor()*0;
            return precio*(1+IVA);
        }
        else if (Objects.equals(getTipoEps(), "Contributivo"))
        {
            precio=(habitacionAsignada.getCategoria().getValor()/2)*habitacionAsignada.getDias();
            return precio*(1+IVA);
        }
        else
        {
            precio=habitacionAsignada.getCategoria().getValor()*habitacionAsignada.getDias();
            return precio*(1+IVA);
        }
    }

    /* Método que busca las vacunas por tipo ingresada en el parámetro, invocando el método
     de hospital, y con el for filtrandolas por eps del paciente */
    public ArrayList<Vacuna> buscarVacunaPorEps(String tipo, Hospital hospital){
        ArrayList<Vacuna> vacunasPorTipo= hospital.buscarTipoVacuna(tipo);
        ArrayList<Vacuna> vacunasDisponibles= new ArrayList<Vacuna>();

        for (int i=1; i<=vacunasPorTipo.size();i++){
            ArrayList<String> tipoEpsVacuna = vacunasPorTipo.get(i-1).getTipoEps();
            for (int j=1;j<=tipoEpsVacuna.size();j++){
                if (Objects.equals(tipoEpsVacuna.get(j - 1), getTipoEps())){
                    vacunasDisponibles.add(vacunasPorTipo.get(i-1));
                }
            }
        }
        return vacunasDisponibles;
    }
    //Método para bienvenida de doctor
    public String mensajeDoctor(Persona doctor){
        //Hay ligadura dinámica
        return doctor.bienvenida()+ "\nPor favor selecciona los medicamentos que vas a formularle a: "+getNombre();
    }

    //Getter y setter

    //Agrega una cita de vacuna al historial de vacunas del paciente
    public void actualizarHistorialVacunas(CitaVacuna citaAsignada){
        historiaClinica.getHistorialVacunas().add(citaAsignada);
    }

    public HistoriaClinica getHistoriaClinica() {
        return historiaClinica;
    }

    public CategoriaHabitacion getCategoriaHabitacion()
    {
        return categoriaHabitacion;
    }

    public void setCategoriaHabitacion(CategoriaHabitacion categoriaHabitacion) {
        this.categoriaHabitacion = categoriaHabitacion;
    }

    public Habitacion getHabitacionAsignada() {
        return habitacionAsignada;
    }

    public void setHabitacionAsignada(Habitacion habitacionAsignada) {
        this.habitacionAsignada = habitacionAsignada;
    }

    @Override

    public String toString(){
        return "---------------------------\nNombre: "+ getNombre() + "\nCédula: " + getCedula() + "\nTipo de EPS: " + getTipoEps() + "\n---------------------------";
    }

    public String despedida(Cita citaAsignada ){
        /*Hay ligadura dinámica ya que la referencia
        citaAsignada puede apuntar a una cita o a una cita vacuna
        y dependiendo de eso ejecutará el método más específico
        de mensaje().
         */
        return "¡Adiós "+getNombre()+" "+citaAsignada.mensaje();
    }
}

File: Python/src/gestor_aplicacion/personas/paciente.py
from src.gestor_aplicacion.administracion.historia_clinica import HistoriaClinica
from src.gestor_aplicacion.personas.persona import Persona
from src.gestor_aplicacion.servicios.cita import Cita
from src.gestor_aplicacion.servicios.cita_vacuna import CitaVacuna
from src.gestor_aplicacion.servicios.formula import Formula
from src.gestor_aplicacion.servicios.habitacion import Habitacion
from src.manejo_errores.error_aplicacion import SinDoctores


class Paciente(Persona):

    def __init__(self, cedula, nombre, tipo_eps):
        super().__init__(cedula, nombre, tipo_eps)
        self._habitacion_asignada = None
        self._HISTORIA_CLINICA = HistoriaClinica(self)

    def calcular_precio(self, servicio):
        if isinstance(servicio, Formula):
            IVA = 1.19
            precio_total = 0
            for med in servicio.lista_meds:
                if self._tipo_eps == "Subsidiado":
                    precio_total += med.precio * 1.3
                if self._tipo_eps == "Contributivo":
                    precio_total += med.precio * 1.2
                if self._tipo_eps == "Particular":
                    precio_total += med.precio
            return precio_total * IVA
        elif isinstance(servicio, CitaVacuna):
            IVA = 0.19
            precio_total = servicio.vacuna.precio

            if servicio.vacuna.tipo == "Obligatoria":
                precio_total += 1000
            elif servicio.vacuna.tipo == "No obligatoria":
                precio_total += 3000

            tipo_eps = servicio.paciente.tipo_eps
            if tipo_eps == "Contributivo":
                precio_total += 2000
            elif tipo_eps == "Subsidiado":
                precio_total += 500
            elif tipo_eps == "Particular":
                precio_total += 10000

            precio_total *= (1 + IVA)
            return precio_total
        elif isinstance(servicio, Habitacion):
            IVA = 0.19
            precio_total = 0
            if self.tipo_eps == "Subsidiado":
                precio_total = self._habitacion_asignada.categoria.valor * 0
            elif self.tipo_eps == "Contributivo":
                precio_total = (self._habitacion_asignada.categoria.valor / 2) * self._habitacion_asignada.dias
            else:
                precio_total = self._habitacion_asignada.categoria.valor * self._habitacion_asignada.dias
            precio_total *= (1 + IVA)
            return precio_total

        elif isinstance(servicio, Cita):
            IVA = 0.19
            precio_total = 0
            tipo_eps = servicio.paciente.tipo_eps
            especialidad = servicio.doctor.especialidad

            if especialidad == "General":
                precio_total += 5000
            if especialidad == "Odontologia":
                precio_total += 10000
            if especialidad == "Oftalmologia":
                precio_total += 7000
            if tipo_eps == "Contributivo":
                precio_total += 2000
            if tipo_eps == "Subsidiado":
                precio_total += 500
            if tipo_eps == "Particular":
                precio_total += 10000

            precio_total *= (1 + IVA)
            return precio_total

    def med_enfermedad(self, enfermedad, hospital):
        meds = hospital.meds_disponibles()
        med_enfermedades = []
        for med in meds:
            if med.enfermedad.nombre == enfermedad.nombre and med.enfermedad.tipologia == enfermedad.tipologia:
                med_enfermedades.append(med)
        return med_enfermedades

    def buscar_doctor_por_eps(self, especialidad, hospital):
        doctores_por_especialidad = hospital.buscar_tipo_doctor(especialidad)
        doctores_disponibles = []

        for doctor in doctores_por_especialidad:
            if doctor.tipo_eps == self.tipo_eps:
                doctores_disponibles.append(doctor)
        if len(doctores_disponibles) != 0:
            return doctores_disponibles
        else:
            raise SinDoctores()


    def buscar_vacuna_por_eps(self, tipo, hospital):
        vacunas_por_tipo = hospital.buscar_tipo_vacuna(tipo)
        vacunas_disponibles = []

        for vacuna in vacunas_por_tipo:
            for eps in vacuna.tipo_eps:
                if eps == self.tipo_eps:
                    vacunas_disponibles.append(vacuna)
        return vacunas_disponibles

    def actualizar_historial_citas(self, cita_seleccionada):
        self.historia_clinica.historial_citas.append(cita_seleccionada)

    def actualizar_historial_vacunas(self,cita_asignada):
        self.historia_clinica.historial_vacunas.append(cita_asignada)

    @property
    def habitacion_asignada(self):
        return self._habitacion_asignada

    @property
    def historia_clinica(self):
        return self._HISTORIA_CLINICA

    @habitacion_asignada.setter
    def habitacion_asignada(self, hab):
        self._habitacion_asignada = hab
